[DRAFT] DOCS:Next.js (APP ROUTER) (#4)
StructCE/DocStruct: stack/next.js/guia(app).md

==- Exemplo de client side

```tsx page.tsx
"use client"

import { getMensagem, postMensagem } from "../../clientApi/mensagem.ts"
import { useState } from "react";

export default function Page() {
  const [mensagem, setMensagem] = useState("");
  const [novaMensagem, setNovaMensagem] = useState("");

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
		setNovaMensagem(e.target.value);
	}

	function handleSubmit(e: FormEvent<HTMLFormElement>) {
		e.preventDefault();
		postMensagem(novaMensagem);
	}

  return (
  // Uso da api junto com useState
		<div>
			<p>Mensagem client-side: {mensagem}</p>
			<button
				type="button"
				onClick={async () => {
					await getMensagem().then((data) => {
						if (data) setMensagem(data.mensagem.texto);
					});
				}
			>
				Busque uma mensagem.
			</button>)

// Form para uso do POST
      <form onSubmit={handleSubmit}>
				<input
					onChange={handleChange}
					placeholder="Nova mensagem"
					name="mensagem"
				/>
				<button
					type="submit"
				>
					Criar mensagem
				</button>
			</form>
    </div>
}
```

===

## Interações via Server Actions

> Outra forma interessante de realizar interações com o servidor é a `Server Action`, com ela é possível criar componentes que fazem chamadas ao servidor para realizar alguma ação. Essas ações podem ser mudanças no banco de dados com validação feita no servidor, por exemplo.

Reviewer: ```suggestion
Outra forma interessante de realizar interações com o servidor é a `Server Action`. Com ela é possível criar componentes que fazem chamadas ao servidor para realizar alguma ação. Essas ações podem ser mudanças no banco de dados com validação feita no servidor, por exemplo.
```

### Criando uma Server Action

> Para criar uma `Server Action` é necessário inserir a diretiva `"use server"` no início de uma função assíncrona, nela será inserida a lógica que rodará no servidor. Aqui será explicada utilizando `Forms`, se quiser aprender utilizando `useEffect` ou `event handlers` entre [aqui](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#non-form-elements).
> 
> Para conseguir acessar os dados do ´Forms´, deverá ser adicionado um parâmetro `formData` do tipo `FormData` na função que define a `server action`. Depois, na tag `<form action={ServerAction}>` do forms adicione a função criada no atributo `action`.

Reviewer: ```suggestion
Para criar uma `Server Action`, é necessário inserir a diretiva `"use server"` no início de uma função assíncrona. Nela, será inserida a lógica que rodará no servidor. Aqui será explicada utilizando `Forms`. Se quiser aprender utilizando `useEffect` ou `event handlers` entre [aqui](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#non-form-elements).

Para conseguir acessar os dados do `Forms`, deverá ser adicionado um parâmetro `formData` do tipo `FormData` na função que define a `server action`. Depois, na tag `<form action={ServerAction}>` do forms, adicione a função criada no atributo `action`.
```

	return (
		<div>
			<h2 className="text-black font-light">Mensagens</h2>
			{mensagens.map((msg) => {
				return (
					<form action={excluirMensagem} key={msg.id}>
						<input hidden name="id" value={msg.id} />
						<p className="text-black font-light">{msg.texto}</p>
						<button className="bg-slate-300 rounded-xl text-black disable:bg-slate-600">
							Excluir
						</button>
					</form>
				);
			})} {/* Inputs com tipo "hidden" podem ser utilizados para passar alguma
			informação para a server action. */}
			<div>
				<h2>Mensagens</h2>
				{mensagens.map((msg) => {
					return <p key={msg.id}>{msg.texto}</p>;
				})}
			</div>
			// Forms recebe a Server Action
			<form action={criarMensagem}>
				<h3>Crie uma mensagem</h3>
				<input name="texto" placeholder="Mensagem" required />
				<button>Enviar</button>
			</form>
		</div>
	);
}
```

!!!
A função `revalidatePath` possui uma interação interessante com o uso de `Server Actions`, quando juntas elas podem criar certo grau de interatividade na página.
!!!

!!!
O uso de um `input` com atributo `hidden` pode ser útil quando precisar de passar alguma informação para a `server action` que não será fornecida pelo usuário.

> !!!

Reviewer: ```suggestion
A função `revalidatePath` possui uma interação interessante com o uso de `Server Actions`. Quando juntas elas podem criar certo grau de interatividade na página.
!!!

!!!
O uso de um `input` com atributo `hidden` pode ser útil quando precisar passar alguma informação para a `server action` que não será fornecida pelo usuário.
!!!
```

## Renderização

A renderização de aplicações next ocorrem tanto em `server-side` quanto em `client-side`, desse modo, a UI criada pode ser beneficiada com `caching` e `interatividade`, por exemplo.

### Server Components

> Os `server components` são utilizados para renderização de UI no lado do servidor e podem sofrer `cache`.

Reviewer: ```suggestion
Os `server components` são utilizados para renderização de UI no lado do servidor e podem ser beneficiados com `caching`.
```

#### Benefícios

-   Interatividade pode ser construída com a utilização de `states`, `effects` e `event listeners`.

-   É possível utilizar APIs de browser.

#### Criando um Client Component

Para criar um `client component` é necessária a presença da diretiva `"use client"` no ínicio do arquivo. Desse modo, os módulos importados irão fazer parte do `client bundle`. Sem a diretiva, ao tentar utilizar um `hook`, por exemplo, ocorrerá um erro.

### Padrões de uso

=== Server components

-   Busca de dados externos

-   Acesso direto ao banco de dados

-   Utilização de dados sensíveis (Chaves de API, Tokens, ...)

===

=== Client components

-   Uso de `event listeners` (`onClick`, `onChange`)

-   Uso de `state` e `effect` (`useState`, `useEffect`)

-   Uso de APIs do navegador

===

## Caching

!!!
Tópico não necessário para a produção de aplicações next.

O next apresenta diferentes tipos de `cache` que poderão ser vistos na [documentação](https://nextjs.org/docs/app/building-your-application/caching).

Se o cache for mal feito, a página pode conter erros como dados desatualizados.
!!!

> Por padrão, o next coloca pedaços estáticos de páginas e resultados de `fetch` no `cache`, contudo, é possível manipular se algo irá para `cache` ou não, quanto tempo permancerá em `cache` e até atualizar o `cache` após certo evento. Para isso, poderão ser utilizados diferentes métodos como `cache()`, `revalidatePath()` e configurações de rota, por exemplo.

Reviewer: ```suggestion
Por padrão, o next coloca pedaços estáticos de páginas e resultados de `fetch` no `cache`. Contudo, é possível manipular se algo irá para `cache` ou não, quanto tempo permancerá em `cache` e até atualizar o `cache` após certo evento. Para isso, poderão ser utilizados diferentes métodos como `cache()`, `revalidatePath()` e configurações de rota, por exemplo.
```

### Server Cache

#### Fetch

> Resultados recebidos pelo `fetch`, quando utilizado `server side` são memorizados até que o componente seja renderizado, por padrão, mas podem ser ajustados passando mais parâmetros para a função.

Reviewer: ```suggestion
Resultados recebidos pelo `fetch`, quando utilizado o `server side`, são memorizados até que o componente seja renderizado, por padrão, mas podem ser ajustados passando mais parâmetros para a função.
```

É possível não gerar cache utilizando:

```tsx page.tsx
fetch("url", { cache: "no-store" });
```

É possível "setar" um tempo, em segundos, para revalidar a informação utilizando:

```tsx page.tsx
fetch("url", { next: { revalidate: 3600 } }); // Revalida a cada 1 hora
```

### React cache

Se alguma informação faz sentido ser memorizada, mas não é automaticamente, ela pode ir pra cache utilizando a função `cache` do react.

Abaixo tem um exemplo com uma busca no banco de dados:

```ts getItens.ts
import { cache }
import prisma from "../prisma/index"

export const getItens = cache(async () => {
	const itens = await prisma.item.findMany()
	return itens
})
```

### revalidatePath

> A função revalidatePath é bem interessante quando alguma rota precisa de sofrer alguma atualização devido um evento.

Reviewer: ```suggestion
A função `revalidatePath` é bem interessante quando alguma rota precisa sofrer alguma atualização devido a um evento.
```

Exemplo com server action:

```tsx page.tsx
export default async function Page() {
	const criarItem = async (formData: FormData) => {
		// Lógica por trás

		revalidatePath("path");
	};

	const itens = prisma.item.findMany();

	// return da página
}
```

### Configurações de rota

> Uma rota estática pode ser forçada a ser dinâmica ou pode receber um tempo entre revalidações recebendo "atributos". Para forçar a página a ser dinâmica, é preciso colocar a linha `export const dynamic = "force-dynamic";` e para revalidar é possível adicionando `export const revalidate = 3600; // tempo em segundos`.

Reviewer: ```suggestion
Uma rota estática pode ser forçada a ser dinâmica ou pode receber um tempo entre revalidações recebendo "atributos". Para forçar a página a ser dinâmica, é preciso colocar a linha `export const dynamic = "force-dynamic";` e para revalidar, é necessário adicionar `export const revalidate = 3600; // tempo em segundos`.
```

### Client cache

> O next cria um `cache` de rotas no navegador do usuário. Se esse `cache` estiver atrapalhando alguma funcionalidade, poderá ser utilizada a função `router.refresh()` do hook `useRouter`.

Reviewer: Eu acho que cabe colocar um link pra documentação aqui.

```suggestion
O next cria um `cache` de rotas no navegador do usuário. Se esse `cache` estiver atrapalhando alguma funcionalidade, poderá ser utilizada a função [`router.refresh()` do hook `useRouter`](https://nextjs.org/docs/app/api-reference/functions/use-router#userouter).
```

> ## Optimizações
> 
> O next apresenta optimizações `built-in` para melhorar a velocidade da aplicação e as [métricas da web](https://web.dev/articles/vitals?hl=pt-br). As `Images`, `Fonts` e `Metadata` serão abordadas nesse tópico, para mais recursos de optimização acesse [optimizações](https://nextjs.org/docs/app/building-your-application/optimizing).

Reviewer: ```suggestion
## Otimizações

O next apresenta otimizações `built-in` para melhorar a velocidade da aplicação e as [métricas da web](https://web.dev/articles/vitals?hl=pt-br). As `Images`, `Fonts` e `Metadata` serão abordadas nesse tópico. Para mais recursos de otimização, acesse [*optimizing*](https://nextjs.org/docs/app/building-your-application/optimizing).
```

Author: acho que aqui seria optimização mesmo porque seria uma parte dedicada a melhorias.

Reviewer: Certeza? Otimização é justamente isso, não sei se optimização é certo (o corretor automático até reclama aqui)

Author: Eu pesquisei optimização no google e tinha aparecido algo, mas eu vi aqui de novo e realmente parece que não existe vou trocar aqui.

### Imagens

> O next tem um componente chamado `Image` que extende a funcionalidade da `img` do html. Esse componente optimiza

Reviewer: ```suggestion
O next tem um componente chamado `Image` que extende a funcionalidade da `img` do html. Esse componente otimiza
```

o tamanho do arquivo de imagem, promove estabilidade visual e utiliza `lazy loading`.

> O componente automaticamente converte a imagem para formatos modernos de imagem como `WebP` e `ÀVIF`.

Reviewer: ```suggestion
O componente automaticamente converte a imagem para formatos modernos de imagem como `WebP` e `AVIF`.
```

            // height={500}
            // blurDataURL="data:..."
            // placeholder="blur" //
		/>
	);
}
```

#### Imagens remotas

Para imagens armazenadas remotamente, será necessário preencher os campos de `src`, `alt`, `width` e `height`. Além disso, o next precisa de uma modificação no arquivo `next.config.mjs` para permitir o uso de imagens externas.

A `src` deve ser preenchida com a string de uma URL.

```js next.config.mjs
/** @type {import('next').NextConfig} */
const nextConfig = {
	images: {
		remotePatterns: [
			{
				protocol: "https",
				hostname: "Nome do host",
				port: "Se necessário, especificar a porta",
				pathname: "Caminho até a imagem",
			},
		],
	},
};

export default nextConfig;
```

```tsx page.tsx
import Image from "next/image";

export default function Page() {
	return (
		<Image
			src="alguma_url"
			alt="Texto alternativo"
			width={500}
			height={500}
		/>
	);
}
```

!!!
Para gerar prioridade a uma imagem, adicione a propriedade `priority` a ela.
!!!

### Fontes

!!!
Se estiver usando tailwind no projeto, acesse [com TailWind](https://nextjs.org/docs/app/building-your-application/optimizing/fonts#with-tailwind-css)
!!!

> As `google fonts` podem ser importadas do next e utilizadas sem nenhuma configuração em outra arquivo.

Reviewer: ```suggestion
As `google fonts` podem ser importadas no next e utilizadas sem nenhuma configuração em outro arquivo.
```
